src/lib/tpm.c

```diff
@@ -486,6 +499,77 @@ bool files_load_bytes_from_path(const char *path, UINT8 *buf, UINT16 *size) {
 LOAD_TYPE(TPM2B_PUBLIC, public)
 LOAD_TYPE(TPM2B_PRIVATE, private)
 
+CK_RV tpm_get_token_info_from_tpm (tpm_ctx *ctx, CK_TOKEN_INFO *info) {
+
+    uint32_t revision;
+    uint32_t version;
+    uint32_t manufacturer;
+    unsigned char manufacturerID[5] = {0}; // 4 bytes + '\0'
+    uint32_t vendor[4];
+    CK_RV rv = CKR_OK;
+    TPM2_CAP capability = TPM2_CAP_TPM_PROPERTIES;
+    UINT32 property = TPM2_PT_FIXED;
+    UINT32 propertyCount = TPM2_MAX_TPM_PROPERTIES;
+    TPMS_CAPABILITY_DATA *capabilityData;
+    TPMI_YES_NO moreData;
+
+    check_pointer(ctx);
+    check_pointer(info);
+
+    TSS2_RC rval = Esys_GetCapability(ctx->esys_ctx,
+        ESYS_TR_NONE,
+        ESYS_TR_NONE,
+        ESYS_TR_NONE,
+        capability,
+        property, propertyCount, &moreData, &capabilityData);
+
+    if (rval != TSS2_RC_SUCCESS) {
+        LOGE("Esys_GetCapability: 0x%x:", rval);
+        return CKR_GENERAL_ERROR;
+    }
+
+    if (!capabilityData || capabilityData->data.tpmProperties.count < TPM2_PT_VENDOR_STRING_4 - TPM2_PT_FIXED + 1) {
+        LOGE("TPM did not reply with correct amount of capabilities");
+	rv = CKR_GENERAL_ERROR;
+	goto out;
+    }
+
+    // Use Spec revision as HW Version
+    revision = capabilityData->data.tpmProperties.tpmProperty[TPM2_PT_REVISION - TPM2_PT_FIXED].value;
+    info->hardwareVersion.major = revision / 100;
+    info->hardwareVersion.minor = revision % 100;
+
+    // Use Firmware Version as FW Version
+    version = capabilityData->data.tpmProperties.tpmProperty[TPM2_PT_FIRMWARE_VERSION_1 - TPM2_PT_FIXED].value;
+    // Most vendors seem to use 00MM.00mm as format for TPM2_PT_FIRMWARE_VERSION_1
+    // Unfortunately we only have 1 byte for major and minor each.
+    info->firmwareVersion.major = (version >> 16) & 0xFF;
+    info->firmwareVersion.minor = version  & 0xFF;
+
+    // Use Vendor String as Model
+    manufacturer = ntohl(capabilityData->data.tpmProperties.tpmProperty[TPM2_PT_MANUFACTURER - TPM2_PT_FIXED].value);
+    memcpy(manufacturerID, (unsigned char*) &manufacturer, sizeof(uint32_t));
+    str_padded_copy(info->manufacturerID, manufacturerID, sizeof(info->manufacturerID));
+
+    // Map human readable Manufacturer String, if available
+    for (unsigned int i=0; i < ARRAY_LEN(TPM2_MANUFACTURER_MAP); i++){
+        if (strncasecmp((char *)info->manufacturerID, (char *)TPM2_MANUFACTURER_MAP[i][0], 4) == 0){
+            str_padded_copy(info->manufacturerID, (unsigned char *)TPM2_MANUFACTURER_MAP[i][1], sizeof(info->manufacturerID));
```

Comment:
if we do this, we need a default of just putting the bytes in there.

Reply:
Which is exactly done in the memcpy + str_padded_copy in the lines above. (551f)
So only if there is a match in the table, the bytes will be modified, otherwise we just have the short hand identifier padded with blanks.